### PR TITLE
migrate to go 1.14 + modules

### LINKED ISSUE #1886: Migrate to go modules
As of Go v1.13, [modules](https://blog.golang.org/go1.13) are the default dependency management tool.

Upstream dependencies like [apimachinery](https://github.com/kubernetes/apimachinery/blob/master/go.mod), [client-go](https://github.com/kubernetes/client-go/blob/master/go.mod), and the [AWS SDK](https://github.com/aws/aws-sdk-go-v2/blob/master/go.mod) have moved.

Velero should adopt go modules as well in order to keep moving forward with the ecosystem.

### PR DESCRIPTION
**Pull request by skriss**
closes #1886 
xref #2127 

This PR migrates Velero from `dep` to go modules for dependency management, and upgrades the go version to ~1.13~ 1.14. Heavily inspired by @ashish-amarnath's work in #2127.

Probably easiest to review commit-by-commit, as I broke it up into small steps.

I *believe* the only thing that's not using modules yet is the Kubernetes `code-generator` tool - it's still being cloned into a $GOPATH-based dir and being run from there due to some compatibility issues (I'll update with links to issues).

I've tested all of the main workflows and things seemed to be working properly, but I would definitely appreciate some further testing before getting ready to merge this.

### CODE REVIEW

**Comment by carlisia** [comment]
Hey @skriss I pulled this branch locally to run tests. Got some errors "not finding package" and am thinking there needs to be a go mod command run? I can google but should some documentation be added to the `Contribute` section?

**Comment by skriss** [comment]
> Hey @skriss I pulled this branch locally to run tests. Got some errors "not finding package" and am thinking there needs to be a go mod command run? I can google but if that's the case, should some documentation be added to the Contribute section?

Hmm, not sure off the top of my head what issue you're running into -- can you provide some more detail (go version, is the code in `$GOPATH`, what's `$GO111MODULE` set to, what command did you run, full error output)?

**Comment by carlisia** [comment]
Nvm, all I had to do was upgrade to 1.13.

File: Makefile [hunk 1]
```diff
@@ -127,21 +127,22 @@ BUILDER_IMAGE := velero-builder
 
 # Example: make shell CMD="date > datefile"
```
Comment (ashish-amarnath) [inline Makefile]: We want to have a make target to verify that the `go.[mod sum]` don't go out of date.

File: hack/build-image/Dockerfile [hunk 2]
```diff
@@ -12,40 +12,38 @@
...
 RUN git config --global advice.detachedHead false
 RUN git clone -b kubernetes-1.17.0 https://github.com/kubernetes/code-generator
-RUN git clone -b kubernetes-1.17.0 https://github.com/kubernetes/apimachinery 
-# Use a proxy for go modules to reduce the likelihood of various hosts being down and breaking the build
-ENV GOPROXY=https://proxy.golang.org
 # vendor code-generator go modules to be compatible with pre-1.15
-WORKDIR  /go/src/k8s.io/code-generator
+WORKDIR /go/src/k8s.io/code-generator
 # Don't use ENV here because we don't want to disable modules for subsequent commands
 RUN GO111MODULE=on go mod vendor
```
Comment (ashish-amarnath) [inline hack/build-image/Dockerfile]: Should we change this to use the env var now that we are cutting over to go modules?

File: hack/update-generated-crd-code.sh [hunk 3]
```diff
@@ -28,8 +30,8 @@ if [[ ! -d "${GOPATH}/src/k8s.io/code-generator" ]]; then
   exit 1
 fi
 
-if [[ ! -d "${GOPATH}/src/sigs.k8s.io/controller-tools" ]]; then
-  echo "sigs.k8s.io/controller-tools missing from GOPATH"
+if ! command -v controller-gen > /dev/null; then
+  echo "controller-gen is missing"
```
Comment (ashish-amarnath) [inline hack/update-generated-crd-code.sh]: should we also say how to get it?

File: Makefile (hunk 1 above)
Comment (ashish-amarnath) [reply Makefile]: https://github.com/vmware-tanzu/velero/pull/2144/files#diff-b67911656ef5d18c4ae36cb6741b7965R295

File: hack/build-image/Dockerfile (hunk 2 above)
Comment (ashish-amarnath) [inline hack/build-image/Dockerfile]: `go mod download` seems to be the newer way to do this.
https://github.com/golang/go/wiki/Modules#how-do-i-use-vendoring-with-modules-is-vendoring-going-away

File: hack/build-image/Dockerfile [hunk 4]
```diff
@@ -12,40 +12,38 @@
...
-WORKDIR /go/src/sigs.k8s.io/controller-tools
-RUN GO111MODULE=on go mod vendor
-RUN go get golang.org/x/tools/cmd/goimports
-WORKDIR /go/src/golang.org/x/tools
-RUN git checkout 40a48ad93fbe707101afb2099b738471f70594ec
-RUN go install ./cmd/goimports
-RUN echo chmod -R a+w /go
+
+# get controller-tools
+RUN GO111MODULE=on go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.2.2
```
Comment (ashish-amarnath) [inline hack/build-image/Dockerfile]: how will we track/maintain this specific dependency?

**Comment by skriss** [comment]
moving out of draft but I still have some followups

**Comment by nrb** [comment]
Putting this on hold so it can be moved to post-1.3.

File: hack/build-image/Dockerfile (hunk 2 above)
Comment (skriss) [reply hack/build-image/Dockerfile]: turns out we don't actually need to do this at all (as far as I can tell), removing.

File: hack/build-image/Dockerfile (hunk 4 above)
Comment (skriss) [reply hack/build-image/Dockerfile]: for now we just track it through the Dockerfile. I'm going to file a follow-up issue to look at migrating to a `tools.go` approach.

File: Makefile (hunk 1 above)
Comment (skriss) [reply Makefile]: :+1: I'll borrow from the CSI plugin.

File: hack/update-generated-crd-code.sh (hunk 3 above)
Comment (skriss) [reply hack/update-generated-crd-code.sh]: this is really only intended to be run inside the build container, which already includes it, so I don't think this matters too much

**Comment by skriss** [comment]
rebased, addressed feedback, and updated to go 1.14. let's see if it passes CI..

**Comment by ashish-amarnath** [comment]
The CI is passing 🎉

**Comment by skriss** [comment]
I'll do a little bit of retesting on this just to ensure nothing broke with the latest changes, but I think it's good to go.

We'll hold until just after we release 1.3.

File: hack/build-image/Dockerfile [hunk 5]
```diff
@@ -14,6 +14,7 @@
 
 FROM golang:1.13
 
+ENV GO111MODULE=on
```
Comment (nrb) [inline hack/build-image/Dockerfile]: I actually don't know that we need this set to `on` anymore, given it's on by default in 1.14. Did you find that it was necessary as you were testing this out during your changes?

Comment (skriss) [reply hack/build-image/Dockerfile]: probably not needed, let me try things out without it - thanks for catching.

Comment (skriss) [reply hack/build-image/Dockerfile]: looks like it's still needed to do `go get` with an `@<version>` suffix, since we're doing that outside of a module.

Comment (skriss) [reply hack/build-image/Dockerfile]: Specifically, for:

```
RUN go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.2.4
```

File: hack/update-generated-crd-code.sh [hunk 6]
```diff
@@ -38,12 +40,13 @@ ${GOPATH}/src/k8s.io/code-generator/generate-groups.sh \
   github.com/vmware-tanzu/velero/pkg/generated \
   github.com/vmware-tanzu/velero/pkg/apis \
   "velero:v1" \
-  --go-header-file ${GOPATH}/src/github.com/vmware-tanzu/velero/hack/boilerplate.go.txt \
+  --go-header-file ./hack/boilerplate.go.txt \
+  --output-base ../../.. \
   $@
 
-go run ${GOPATH}/src/sigs.k8s.io/controller-tools/cmd/controller-gen/main.go \
+controller-gen \
```
Comment (jenting) [inline hack/update-generated-crd-code.sh]: :heart:

Comment (jenting) [inline hack/update-generated-crd-code.sh]: :heart:

**Comment by skriss** [comment]
I retested the CI steps as well as the local dev stuff (go build/go test locally) inside and out of $GOPATH, and everything LGTM, so from my perspective this is ready to go. I will file a couple follow-up enhancement issues as well.

**Pull request merged**